### PR TITLE
persist build api credentials

### PR DESCRIPTION
**Pull request by jemoreira**

### CODE REVIEW

File: conf.toml [hunk 1]
```diff
@@ -16,6 +16,15 @@ OAuthClientResourceID = ""
 [SecretManager.UNIX]
 SecretFilePath = "../secrets.json"
 
+[EncryptionService]
+Type = "Simple"
+
+[EncryptionService.Simple]
+KeySizeBits = 128
+
+[EncryptionService.GCP_KMS]
```
Comment (ser-io) [inline conf.toml:25]: we declared above the type to be  "Simple", is this config declaration needed?

File: pkg/app/net/gcp/spanner.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	credentialsTable  = "Credentials"
+	usernameColumn    = "username"
+	credentialsColumn = "credentials"
+)
+
+// A database service that works with a Cloud Spanner database with the following schema:
```
Comment (ser-io) [inline pkg/app/net/gcp/spanner.go:42]: Let's create a bug to track setting up the database programmatically where we can check in this schema code.

File: conf.toml (hunk 1 above)
Comment (jemoreira) [reply conf.toml:25]: No, I just included it as a reference

File: pkg/app/types.go [hunk 3]
```diff
@@ -108,3 +108,11 @@ type EncryptionService interface {
 	Encrypt(plaintext []byte) ([]byte, error)
 	Decrypt(ciphertext []byte) ([]byte, error)
 }
+
+type DatabaseService interface {
+	// Credentials are usually stored encrypted hence the []byte type.
+	// If no credentials are available for the given user Fetch returns nil, nil.
+	FetchBuildAPICredentials(username string) ([]byte, error)
+	// Store new credentials or overwrite existing ones for the given user.
+	StoreBuildAPICredentials(username string, credentials []byte) error
```
Comment (ser-io) [inline pkg/app/types.go:125]: nit: InsertOrUpdate instead of Store?

File: pkg/app/types.go [hunk 4]
```diff
@@ -109,10 +112,21 @@ type EncryptionService interface {
...
+}
+
 type DatabaseService interface {
 	// Credentials are usually stored encrypted hence the []byte type.
 	// If no credentials are available for the given user Fetch returns nil, nil.
 	FetchBuildAPICredentials(username string) ([]byte, error)
 	// Store new credentials or overwrite existing ones for the given user.
 	StoreBuildAPICredentials(username string, credentials []byte) error
+	// Create or update a user session.
+	CreateOrUpdateSession(s *Session) error
```
Comment (ser-io) [inline pkg/app/types.go:127]: ideally this object would be immutable, as it's not the case wondering whether we could receive a copy instead so we don't risk the object getting modified before being stored.

File: conf.toml (hunk 1 above)
Comment (ser-io) [reply conf.toml:25]: could it be commented out then? as it gives the impression it's required.

Comment (jemoreira) [reply conf.toml:25]: The same applies to the `InstanceManager.GCP` entry below. So far we have put all values, required or not, in this file to make them easier to discover. I'm ok with commenting out the ones that are not required or have default values, but it should be done in a different PR.

Comment (ser-io) [reply conf.toml:25]: I see, didn't realize we were following this pattern, discard the comment then.

File: pkg/app/types.go (hunk 4 above)
Comment (jemoreira) [reply pkg/app/types.go]: Done

**Review by ser-io — APPROVED** [review]

**Pull request merged**